module/zstd/zstd.c

```diff
@@ -0,0 +1,611 @@
+ */
+static void *
+zstd_mempool_alloc(struct zstd_pool *zstd_mempool, size_t size)
+{
+	int i;
+	struct zstd_pool *pool;
+	struct zstd_kmem *mem = NULL;
+
+	if (!zstd_mempool) {
+		return (NULL);
+	}
+
+	/* Seek for preallocated memory slot and free obsolete slots */
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		pool = &zstd_mempool[i];
+		if (mutex_tryenter(&pool->barrier)) {
+			/*
+			 * Check if objects fits the size, if so we take it and
+			 * update the timestamp.
+			 */
+			if (!mem && pool->mem && size <= pool->size) {
+				pool->timeout = gethrestime_sec() +
+				    ZSTD_POOL_TIMEOUT;
+				mem = pool->mem;
+				continue;
+			}
+
+			/* Free memory if object is older than 2 minutes */
+			if (pool->mem && gethrestime_sec() > pool->timeout) {
+				kmem_free(pool->mem, pool->size);
+				pool->mem = NULL;
+				pool->size = 0;
+				pool->timeout = 0;
+			}
+
+			mutex_exit(&pool->barrier);
+		}
+	}
+
+	if (mem) {
+		return (mem);
+	}
+
+	/* If no preallocated slot was found, try to fill in a new one */
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		pool = &zstd_mempool[i];
+		if (mutex_tryenter(&pool->barrier)) {
+			/* Object is free, try to allocate new one */
+			if (!pool->mem) {
+				mem = vmem_alloc(size, KM_SLEEP);
+				pool->mem = mem;
+
+				/* Allocation successfull? */
+				if (pool->mem) {
+					/* Keep track for later release */
+					mem->pool = pool;
+					pool->size = size;
+					mem->kmem_type = ZSTD_KMEM_POOL;
+					mem->kmem_size = size;
+				}
+			}
+
+			if (size <= pool->size) {
+				/* Update timestamp */
+				pool->timeout = gethrestime_sec() +
+				    ZSTD_POOL_TIMEOUT;
+
+				return (pool->mem);
+			}
+
+			mutex_exit(&pool->barrier);
+		}
+	}
+
+	/*
+	 * If the pool is full or the allocation failed, try lazy allocation
+	 * instead.
+	 */
+	if (!mem) {
+		mem = vmem_alloc(size, KM_NOSLEEP);
+		if (mem) {
+			mem->pool = NULL;
+			mem->kmem_type = ZSTD_KMEM_DEFAULT;
+			mem->kmem_size = size;
+		}
+	}
+
+	return (mem);
+}
+
+/* Mark object as released by releasing the barrier mutex */
+static void
+zstd_mempool_free(struct zstd_kmem *z)
+{
+	mutex_exit(&z->pool->barrier);
+}
+
+/* Convert internal stored zfs level enum to zstd level */
+static enum zio_zstd_levels
+zstd_cookie_to_enum(int32_t level)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(fastlevels); i++) {
+		if (fastlevels[i].cookie == level) {
+			return (fastlevels[i].level);
+		}
+	}
+
+	/* This shouldn't happen. Cause a panic. */
+	printk(KERN_ERR "%s:Invalid ZSTD level encountered: %d",
+	    __func__, level);
+
+	return (ZIO_ZSTD_LEVEL_DEFAULT);
+}
+
+/* Convert zstd_level to internal stored zfs level */
+static int32_t
+zstd_enum_to_cookie(enum zio_zstd_levels elevel)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(fastlevels); i++) {
+		if (fastlevels[i].level == elevel)
+			return (fastlevels[i].cookie);
+	}
+
+	/* This shouldn't happen. Cause a panic. */
+	printk(KERN_ERR "%s:Invalid ZSTD enum level encountered: %d",
+	    __func__, elevel);
+
+	return (3);
```

Comment:
This should be ZIO_ZSTD_LEVEL_DEFAULT rather than the static number 3

Reply:
done, please check again and mark as resolved

Reply:
Checked
This can be marked as resolved by you @c0d3z3r0